### PR TITLE
#92: KPI report

### PR DESCRIPTION
**Pull request by rombautt**

### CODE REVIEW

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java [hunk 1]
```diff
@@ -124,7 +125,7 @@ public void doBuildGraph(final StaplerRequest request, final StaplerResponse res
                         }
 
                         if (kpiReport != null) {
```
Comment (jgheerardyn) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java]: this is always true

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java [hunk 2]
```diff
@@ -124,7 +125,7 @@ public void doBuildGraph(final StaplerRequest request, final StaplerResponse res
                         }
```
Comment (jgheerardyn) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java]: Imo, the else clause in line 124 can be removed

Comment (jgheerardyn) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java]: Moreover, once kpiReport is null, it remains null so you could exit the while clause

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java [hunk 3]
```diff
@@ -111,7 +112,7 @@ public void doBuildGraph(final StaplerRequest request, final StaplerResponse res
                 if (kpiReport != null) {
 
                     Map<Integer, ScoreResult> testResults = new TreeMap<Integer, ScoreResult>();
-                    testResults.put(0, kpiReport.getKpiScore(testName));
+                    testResults.put(kpiReport.getBuildNumber(), kpiReport.getKpiScore(testName));
 
                     int total = 1;
```
Comment (jgheerardyn) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java]: the total variable is never explicitly user any more in the while loop below. maybe cleaner if you move it into a for loop and do not expose the variable outside the loop?

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReport.java [hunk 4]
```diff
@@ -127,8 +134,8 @@ public void setPreviousReport(KPIReport previousReport) {
     public KPIReportGroup getKpiReportGroup(String groupName) {
 
         if (isBlank(groupName)) {
-            // no group selected, take first one
-            return getKpiReportGroups().iterator().next();
+            // no group selected, take first one, if any
+            return CollectionUtils.isNotEmpty(getKpiReportGroups()) ? getKpiReportGroups().iterator().next() : null;
```
Comment (jgheerardyn) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReport.java]: change to 
return hasKpiReport()? ...

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReportService.java [hunk 5]
```diff
@@ -33,6 +33,7 @@
 
     /**
      * Retrieves the @{link KPIReport} for the given build, and links the KPI Reports of the previous builds up to the specified size.
+     * Only builds that succeeded will be considered.
      *
      * @param build  build for which to fetch the KPI Report
      * @param size   total number of builds to chain to the KPI Report of the current build (including current build)
```
Comment (jgheerardyn) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReportService.java]: maybe change javadoc to 
total number of **successful** builds to chain to the KPI Report of the current build (including current build)

**Comment by jgheerardyn** [comment]
reviewed

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java (hunk 2 above)
Comment (rombautt) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java]: True, fixed.

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReport.java (hunk 4 above)
Comment (rombautt) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReport.java]: better idd, fixed.

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReportService.java (hunk 5 above)
Comment (rombautt) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/KPIReportService.java]: fixed

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java (hunk 3 above)
Comment (rombautt) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java]: total variable is still used to control the number of elements to put in the graph.

**Comment by rombautt** [comment]
Applied code review comments, can you merge if ok?

File: plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java (hunk 3 above)
Comment (jgheerardyn) [inline plugins/kpis/src/main/java/io/yields/plugins/kpi/BuildActionResultsDisplay.java]: yes - but not in the while loop between lines 120 and 128 I think...

**Pull request merged**